code/game/objects/structures/tank_dispenser.dm

```diff
@@ -1,117 +1,135 @@
-/obj/structure/dispenser
+/obj/structure/tank_dispenser
 	name = "tank storage unit"
 	desc = "A simple yet bulky storage device for gas tanks. Has room for up to ten oxygen tanks, and ten hydrogen tanks."
 	icon = 'icons/obj/structures/tank_dispenser.dmi'
 	icon_state = "dispenser"
-	density = 1
-	anchored = 1.0
+	density = TRUE
+	anchored = TRUE
+	material = /decl/material/solid/metal/steel
 	tool_interaction_flags = TOOL_INTERACTION_ANCHOR
+	var/list/oxygen_tanks = 10
+	var/list/hydrogen_tanks = 10
 
-	var/oxygentanks = 10
-	var/hydrogentanks = 10
-	var/list/oxytanks = list()	//sorry for the similar var names
-	var/list/hydtanks = list()
+/obj/structure/tank_dispenser/oxygen
+	hydrogen_tanks = 0
 
-/obj/structure/dispenser/oxygen
-	hydrogentanks = 0
+/obj/structure/tank_dispenser/hydrogen
+	oxygen_tanks = 0
 
-/obj/structure/dispenser/hydrogen
-	oxygentanks = 0
+/obj/structure/tank_dispenser/empty
+	hydrogen_tanks = 0
+	oxygen_tanks = 0
 
-/obj/structure/dispenser/Initialize()
+/obj/structure/tank_dispenser/Initialize()
 	. = ..()
+
+	if(isnum(oxygen_tanks) && oxygen_tanks > 0)
+		var/spawn_oxy = oxygen_tanks
+		oxygen_tanks = list()
+		for(var/i in 1 to spawn_oxy)
+			oxygen_tanks += new /obj/item/tank/oxygen(src)
+	else
+		oxygen_tanks = null
+
+	if(isnum(hydrogen_tanks) && hydrogen_tanks > 0)
+		var/spawn_hyd = hydrogen_tanks
+		hydrogen_tanks = list()
+		for(var/i in 1 to spawn_hyd)
+			hydrogen_tanks += new /obj/item/tank/hydrogen(src)
```

Comment:
I wouldn't complain about this if it was oldcode, but you actually changed it to work like this: do we really need to spawn 20 tanks whenever one of these is mapped in or created?